sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/changefeed/implementation/PartitionSupervisorImpl.java

```diff
@@ -35,70 +36,59 @@
 
     private RuntimeException resultException;
 
-    private ExecutorService executorService;
+    private Scheduler scheduler;
 
-    public PartitionSupervisorImpl(Lease lease, ChangeFeedObserver observer, PartitionProcessor processor, LeaseRenewer renewer, ExecutorService executorService) {
+    public PartitionSupervisorImpl(Lease lease, ChangeFeedObserver observer, PartitionProcessor processor, LeaseRenewer renewer, Scheduler scheduler) {
         this.lease = lease;
         this.observer = observer;
         this.processor = processor;
         this.renewer = renewer;
-        this.executorService = executorService;
+        this.scheduler = scheduler;
 
-        if (executorService == null) {
-            this.executorService = Executors.newFixedThreadPool(3);
+        if (scheduler == null) {
+            this.scheduler = Schedulers.elastic();
         }
     }
 
     @Override
     public Mono<Void> run(CancellationToken shutdownToken) {
-        PartitionSupervisorImpl self = this;
         this.resultException = null;
 
-        ChangeFeedObserverContext context = new ChangeFeedObserverContextImpl(self.lease.getLeaseToken());
+        ChangeFeedObserverContext context = new ChangeFeedObserverContextImpl(this.lease.getLeaseToken());
 
-        self.observer.open(context);
+        this.observer.open(context);
 
+        this.processorCancellation = new CancellationTokenSource();
+
+        this.scheduler.schedule(() -> this.processor.run(this.processorCancellation.getToken())
+            .subscribe());
+
+        this.renewerCancellation = new CancellationTokenSource();
+
+        this.scheduler.schedule(() -> this.renewer.run(this.renewerCancellation.getToken())
+            .subscribe());
+
+        return Mono.just(this)
+            .delayElement(Duration.ofMillis(100))
+            .repeat( () -> !shutdownToken.isCancellationRequested() && this.processor.getResultException() == null && this.renewer.getResultException() == null)
```

Comment:
here and elsewhere. You have two concurrent threads.

thread1: does the checking here if there is any failure

thread2: sets the exception.

I think we have a bug here around thread2 seeing thread1 most recent changes. as the resultException is not set atomically by thread2. please fix in all places.

Reply:
Fixed

Reply:
more info here: https://stackoverflow.com/questions/10996669/java-volatile-private-fields-with-getters-and-setters